Cython/Compiler/Nodes.py

```diff
@@ -5078,11 +5078,25 @@ def generate_type_ready_code(entry, code, heap_type_bases=False):
             return
         if entry.visibility != 'extern':
             code.putln("#if CYTHON_COMPILING_IN_LIMITED_API")
+            tuple_temp = code.funcstate.allocate_temp(py_object_type, manage_ref=True)
+            base_type = scope.parent_type.base_type
+            if base_type:
+                for slot in TypeSlots.slot_table:
+                    if slot.slot_name == "tp_base":
+                        code.putln(
+                            "%s = PyTuple_Pack(1, (PyObject *)%s); %s" % (
+                            tuple_temp,
+                            base_type.typeptr_cname,
+                            code.error_goto_if_null(tuple_temp, entry.pos)))
+                        code.put_gotref(tuple_temp)
             code.putln(
-                "%s = PyType_FromSpec(&%s_spec); %s" % (
+                "%s = PyType_FromSpecWithBases(&%s_spec, %s); %s" % (
```

Comment:
`tuple_temp` is already set to `NULL` explicitly. Example code:

```
static int __Pyx_modinit_type_init_code(void) {
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  __Pyx_RefNannySetupContext("__Pyx_modinit_type_init_code", 0);
  /*--- Type init code ---*/
  #if CYTHON_COMPILING_IN_LIMITED_API
  /* __pyx_t_1 allocated (Python object) */
  __pyx_type_9shrubbery_Shrubbery = PyType_FromSpecWithBases(&__pyx_type_9shrubbery_Shrubbery_spec, __pyx_t_1); if (unlikely(!__pyx_type_9shrubbery_Shrubbery)) __PYX_ERR(0, 3, __pyx_L1_error)
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
  /* __pyx_t_1 released */
```

However, you are right, this should be a `XDECREF`. Let me fix that.

Reply:
It's automatically set to `NULL` at the beginning of the C function, yes. But the code section that is generated here "creates" the temp variable and then reads it without first assigning anything to it. Regardless of it being initialised to `NULL`, the code generation function should say that it didn't just forget to assign something, but really meant it to be `NULL`. That will help readers of this code. In the end, we are writing code for people to read and modify. It does not matter if the C compiler needs a femtosecond longer to discard redundant C code.

That being said, why use the temp variable at all if there is no `base_type`? You could just pass `NULL` directly instead, and do the allocate/release dance only at need.

Reply:
I moved this into an `if base_type else`. If there is no base type, then it will just use: `PyType_FromSpec` which calls `PyType_FromSpecWithBases` with a `NULL` bases.